File: bridge/transport.js
import { SerialPort } from 'serialport';
import { v4 as uuidv4 } from 'uuid';
import * as net from 'net';
import WebSocket from 'ws';
import { decode as cobsDecode, encode as cobsEncode } from 'cobs';
import { JupyterMessageProtocol } from './jmp_converter.js';

/**
 * Abstract base transport class
 * Provides common interface for all transport implementations
 */
class BaseTransport {
    constructor() {
        this.messageHandler = null;
        this.controlMessageHandler = null;
        this.isConnected = false;
        this.connectHandler = null;
        this.disconnectHandler = null;
        this.errorHandler = null;
        this.sessionId = uuidv4();
        this.rxBuffer = Buffer.alloc(0);
    }

    async connect() {
        throw new Error('connect() must be implemented by subclass');
    }

    async disconnect() {
        throw new Error('disconnect() must be implemented by subclass');
    }

    async send(message) {
        throw new Error('send() must be implemented by subclass');
    }

    // Event handlers
    onMessage(handler) {
        this.messageHandler = handler;
    }

    onControlMessage(handler) {
        this.controlMessageHandler = handler;
    }

    onConnect(handler) {
        this.connectHandler = handler;
    }

    onDisconnect(handler) {
        this.disconnectHandler = handler;
    }

    onError(handler) {
        this.errorHandler = handler;
    }

    // Helper methods for subclasses
    _handleMessage(message) {
        if (this.messageHandler) {
            this.messageHandler(message);
        }
    }

    _handleControlMessage(message) {
        if (this.controlMessageHandler) {
            this.controlMessageHandler(message);
        }
    }

    _handleConnect() {
        this.isConnected = true;
        if (this.connectHandler) {
            this.connectHandler();
        }
    }

    _handleDisconnect() {
        this.isConnected = false;
        if (this.disconnectHandler) {
            this.disconnectHandler();
        }
    }

    _handleError(error) {
        if (this.errorHandler) {
            this.errorHandler(error);
        }
    }

    /**
     * Common COBS-framed data handler
     * Used by all transport types for uniform message processing
     */
    _handleFramedData(data) {

        this.rxBuffer = Buffer.concat([this.rxBuffer, data]);

        let delimiterIndex;
        while ((delimiterIndex = this.rxBuffer.indexOf(0x00)) !== -1) {
            const encodedFrame = this.rxBuffer.slice(0, delimiterIndex);
            this.rxBuffer = this.rxBuffer.slice(delimiterIndex + 1);

            if (encodedFrame.length === 0) continue;

            //log raw data recieved 
            //console.debug(`[RX encoded, len=${encodedFrame.length}]`, encodedFrame.toString('hex').match(/.{2}/g).join(' '));


            let decoded;
            try {
                decoded = cobsDecode(encodedFrame);
            } catch (error) {
                console.error('COBS decoding failed:', error);
                this._handleError(error);
                continue;
            }

            //log decoded data
            //console.debug(`[RX decoded, len=${decoded.length}]`, decoded.toString('hex').match(/.{2}/g).join(' '));


            try {
                if (decoded.length < 36) {
                    throw new Error(`Decoded data too short (len=${decoded.length})`);
                }

                //sessionid is a 36-bytes prefix added to support routing when multiclient. 
                //it is expected to be stripped out before the packet leaves the transport layer
                const sessionId = decoded.slice(0, 36).toString('utf8');
                const payload = decoded.slice(36);

                const message = JupyterMessageProtocol.binaryToJson(payload);
                console.debug(message);
                message._sessionId = sessionId;

                this._handleMessage(message);
            } catch (error) {
                console.error('Failed to parse decoded message:', error);
                this._handleError(error);
            }
        }
    }

    /**
     * Common method to encode and frame messages
     * Used by all transport types for uniform message sending
     */
    _encodeMessage(message) {
        const jsonMessage = typeof message === 'string' ? JSON.parse(message) : message;
        const binaryMsg = JupyterMessageProtocol.jsonToBinary(jsonMessage);
        const framedPayload = Buffer.concat([Buffer.from(this.sessionId), binaryMsg]);

        //console.debug(`[TX decoded, len=${framedPayload.length}]`, framedPayload.toString('hex').match(/.{2}/g).join(' '));

        const encoded = cobsEncode(framedPayload);
        const packet = Buffer.concat([encoded, Buffer.from([0x00])]);

        //console.debug(`[TX encoded, len=${packet.length}]`, packet.toString('hex').match(/.{2}/g).join(' '));

        return packet;
    }
}

/**
 * WebSocket transport implementation
 * Connects to a WebSocket server for kernel communication
 * Uses COBS framing for binary data like other transports
 */
class WebSocketTransport extends BaseTransport {
    constructor(config) {
        super();
        this.config = config;
        this.websocket = null;
        this.reconnectDelay = config.reconnect_delay || 5000;
    }

    async connect() {
        return new Promise((resolve, reject) => {
            const wsUrl = this.config.websocket_url || 'ws://localhost:8080/kernel';
            console.log(`Connecting to WebSocket server at ${wsUrl}`);

            this.websocket = new WebSocket(wsUrl);

            // Set binary type to handle binary data properly
            this.websocket.binaryType = 'arraybuffer';

            this.websocket.on('open', () => {
                console.log('WebSocket connection established');
                this._handleConnect();
                resolve();
            });

            this.websocket.on('message', (data) => {
                try {
                    // Convert data to Buffer for uniform processing
                    let buffer;
                    if (data instanceof ArrayBuffer) {
                        buffer = Buffer.from(data);
                    } else if (Buffer.isBuffer(data)) {
                        buffer = data;
                    } else if (typeof data === 'string') {
                        buffer = Buffer.from(data);
                    } else {
                        buffer = Buffer.from(data);
                    }

                    // Use common framed data handler
                    this._handleFramedData(buffer);
                } catch (error) {
                    console.error('Error processing WebSocket message:', error);
                    this._handleError(error);
                }
            });

            this.websocket.on('error', (error) => {
                console.error('WebSocket error:', error);
                this._handleError(error);
                reject(error);
            });

            this.websocket.on('close', () => {
                console.log('WebSocket connection closed');
                this._handleDisconnect();

                // Auto-reconnect if not intentionally disconnected
                if (this.isConnected) {
                    setTimeout(() => this.connect().catch(console.error), this.reconnectDelay);
                }
            });
        });
    }

    async disconnect() {
        this.isConnected = false;
        if (this.websocket) {
            this.websocket.close();
            this.websocket = null;
        }
    }

    async send(message) {
        if (this.websocket && this.websocket.readyState === WebSocket.OPEN) {
            try {
                // Use common encoding method
                const packet = this._encodeMessage(message);
                this.websocket.send(packet);
                return true;
            } catch (error) {
                console.error('Failed to encode or send message:', error);
                this._handleError(error);
                return false;
            }
        } else {
            console.error('WebSocket not connected, cannot send message');
            return false;
        }
    }
}

/**
 * TCP Socket transport implementation
 * Connects to a TCP server for kernel communication
 * Uses COBS framing for binary data like other transports
 */
class SocketTransport extends BaseTransport {
    constructor(config) {
        super();
        this.config = config;
        this.socket = null;
        this.reconnectDelay = config.reconnect_delay || 5000;
    }

    async connect() {
        return new Promise((resolve, reject) => {
            const host = this.config.host || 'localhost';
            const port = this.config.transport_port || 5555;

            console.log(`Connecting to TCP server at ${host}:${port}`);

            this.socket = new net.Socket();

            this.socket.connect(port, host, () => {
                console.log('TCP socket connection established');
                this._handleConnect();
                resolve();
            });

            this.socket.on('data', (data) => {
                // Use common framed data handler
                this._handleFramedData(data);
            });

            this.socket.on('error', (error) => {
                console.error('TCP socket error:', error);
                this._handleError(error);
                reject(error);
            });

            this.socket.on('close', () => {
                console.log('TCP socket connection closed');
                this._handleDisconnect();

                // Auto-reconnect if not intentionally disconnected
                if (this.isConnected) {
                    setTimeout(() => this.connect().catch(console.error), this.reconnectDelay);
                }
            });
        });
    }

    async disconnect() {
        this.isConnected = false;
        if (this.socket) {
            this.socket.destroy();
            this.socket = null;
        }
    }

    async send(message) {
        if (!this.socket || this.socket.destroyed) {
            console.log('TCP socket not connected, attempting to connect...');
            try {
                await this.connect();
            } catch (error) {
                console.error('Failed to connect before sending message:', error);
                return false;
            }
        }

        try {
            // Use common encoding method
            const packet = this._encodeMessage(message);
            this.socket.write(packet);
            return true;
        } catch (error) {
            console.error('Failed to encode or send message:', error);
            this._handleError(error);
            return false;
        }
    }
}

/**
 * Serial transport implementation
 * Communicates over serial port with COBS framing
 */
class SerialTransport extends BaseTransport {
    constructor(config) {
        super();
        this.config = config;
        this.port = null;
    }

    async connect() {
        return new Promise((resolve, reject) => {
            this.port = new SerialPort({
                path: this.config.port_path || '/dev/ttyUSB0',
                baudRate: this.config.baud_rate || 115200,
            });

            this.port.on('open', () => {
                console.log('Serial port opened');
                this._handleConnect();
                resolve();
            });

            this.port.on('data', (data) => {
                // Use common framed data handler
                this._handleFramedData(data);
            });

            this.port.on('error', (error) => {
                console.error('Serial port error:', error);
                this._handleError(error);
                reject(error);
            });

            this.port.on('close', () => {
                console.log('Serial port closed');
                this._handleDisconnect();
            });
        });
    }

    async disconnect() {
        this.isConnected = false;
        if (this.port) {
            this.port.close();
            this.port = null;
        }
    }

    async send(message) {
        if (this.port && this.port.isOpen) {
            try {
                // Use common encoding method
                const packet = this._encodeMessage(message);
                this.port.write(packet);
                return true;
            } catch (error) {
                console.error('Failed to encode or send message:', error);
                this._handleError(error);
                return false;
            }
        } else {
            console.error('Serial port not connected, cannot send message');
            return false;
        }
    }
}

/**
 * TransportFactory creates transport instances based on type
 */
class TransportFactory {
    static create(type, config) {
        switch (type.toLowerCase()) {
            case 'websocket':
            case 'ws':
                return new WebSocketTransport(config);
            case 'socket':
            case 'tcp':
                return new SocketTransport(config);
            case 'serial':
                return new SerialTransport(config);
            default:
                throw new Error(`Unknown transport type: ${type}`);
        }
    }

    static getSupportedTypes() {
        return ['websocket', 'ws', 'socket', 'tcp', 'serial'];
    }
}

export {
    BaseTransport,
    WebSocketTransport,
    SocketTransport,
    SerialTransport,
    TransportFactory
};

export default TransportFactory;

File: ukernel/ukernel/transport.py
# transport layer — connects to the firmware over tcp, serial, or websocket.
# all transports use cobs framing with a 36-byte session-id prefix.

import asyncio
import logging
import uuid
from abc import ABC, abstractmethod

from cobs import cobs

from .protocol import encode_message, decode_message

log = logging.getLogger(__name__)

SESSION_ID_LEN = 36


class Transport(ABC):
    def __init__(self):
        self.session_id = str(uuid.uuid4()).encode("utf-8")[:SESSION_ID_LEN]
        self._on_message = None
        self._on_disconnect = None
        self.connected = False

    def on_message(self, callback):
        self._on_message = callback

    def on_disconnect(self, callback):
        """Called once when the read loop ends (EOF, error, or explicit close).
        Lets the kernel fail any in-flight requests instead of hanging."""
        self._on_disconnect = callback

    def _dispatch_disconnect(self, reason: str):
        if self._on_disconnect:
            try:
                self._on_disconnect(reason)
            except Exception as e:
                log.error("on_disconnect callback raised: %s", e)
            # one-shot — clear so a subsequent reconnect re-registers cleanly
            self._on_disconnect = None

    @abstractmethod
    async def connect(self): ...

    @abstractmethod
    async def disconnect(self): ...

    @abstractmethod
    async def send(self, msg: dict) -> bool: ...

    def _frame_encode(self, msg: dict) -> bytes:
        binary = encode_message(msg)
        payload = self.session_id + binary
        return cobs.encode(payload) + b"\x00"

    def _frame_decode(self, frame: bytes) -> dict:
        decoded = cobs.decode(frame)
        if len(decoded) < SESSION_ID_LEN:
            raise ValueError(f"frame too short ({len(decoded)} bytes)")
        payload = decoded[SESSION_ID_LEN:]
        msg = decode_message(payload)
        msg["_session_id"] = decoded[:SESSION_ID_LEN].decode("utf-8", errors="replace")
        return msg

    def _dispatch(self, msg):
        if self._on_message:
            self._on_message(msg)


class TCPTransport(Transport):
    def __init__(self, host="localhost", port=5555):
        super().__init__()
        self.host = host
        self.port = port
        self._reader: asyncio.StreamReader | None = None
        self._writer: asyncio.StreamWriter | None = None
        self._read_task: asyncio.Task | None = None

    async def connect(self):
        self._reader, self._writer = await asyncio.open_connection(self.host, self.port)
        self.connected = True
        self._read_task = asyncio.create_task(self._read_loop())
        log.info("tcp connected to %s:%d", self.host, self.port)

    async def disconnect(self):
        self.connected = False
        if self._read_task:
            self._read_task.cancel()
            try:
                await self._read_task
            except asyncio.CancelledError:
                pass
        if self._writer:
            self._writer.close()
            try:
                await self._writer.wait_closed()
            except Exception:
                pass
        self._reader = self._writer = None
        log.info("tcp disconnected")

    async def send(self, msg: dict) -> bool:
        if not self._writer:
            return False
        try:
            self._writer.write(self._frame_encode(msg))
            await self._writer.drain()
            return True
        except Exception as e:
            log.error("tcp send failed: %s", e)
            return False

    async def _read_loop(self):
        buf = b""
        reason = "tcp eof"
        try:
            while self.connected:
                chunk = await self._reader.read(4096)
                if not chunk:
                    break
                buf += chunk
                while b"\x00" in buf:
                    idx = buf.index(b"\x00")
                    frame = buf[:idx]
                    buf = buf[idx + 1:]
                    if not frame:
                        continue
                    try:
                        msg = self._frame_decode(frame)
                        self._dispatch(msg)
                    except Exception as e:
                        log.error("frame decode error: %s", e)
        except asyncio.CancelledError:
            reason = "tcp closed"
        except Exception as e:
            reason = f"tcp read error: {e}"
            if self.connected:
                log.error("tcp read error: %s", e)
        finally:
            self.connected = False
            self._dispatch_disconnect(reason)


class SerialTransport(Transport):
    def __init__(self, port_path="/dev/ttyUSB0", baud_rate=115200):
        super().__init__()
        self.port_path = port_path
        self.baud_rate = baud_rate
        self._serial = None
        self._read_task: asyncio.Task | None = None

    async def connect(self):
        import serial
        self._serial = serial.Serial(self.port_path, self.baud_rate, timeout=0)
        self.connected = True
        self._read_task = asyncio.create_task(self._read_loop())
        log.info("serial connected to %s @ %d", self.port_path, self.baud_rate)

    async def disconnect(self):
        self.connected = False
        if self._read_task:
            self._read_task.cancel()
            try:
                await self._read_task
            except asyncio.CancelledError:
                pass
        if self._serial:
            self._serial.close()
            self._serial = None
        log.info("serial disconnected")

    async def send(self, msg: dict) -> bool:
        if not self._serial or not self._serial.is_open:
            return False
        try:
            self._serial.write(self._frame_encode(msg))
            return True
        except Exception as e:
            log.error("serial send failed: %s", e)
            return False

    async def _read_loop(self):
        buf = b""
        loop = asyncio.get_event_loop()
        reason = "serial closed"
        try:
            while self.connected:
                # non-blocking read via executor to avoid blocking the event loop
                data = await loop.run_in_executor(None, self._serial.read, 4096)
                if not data:
                    await asyncio.sleep(0.01)
                    continue
                buf += data
                while b"\x00" in buf:
                    idx = buf.index(b"\x00")
                    frame = buf[:idx]
                    buf = buf[idx + 1:]
                    if not frame:
                        continue
                    try:
                        msg = self._frame_decode(frame)
                        self._dispatch(msg)
                    except Exception as e:
                        log.error("frame decode error: %s", e)
        except asyncio.CancelledError:
            pass
        except Exception as e:
            reason = f"serial read error: {e}"
            if self.connected:
                log.error("serial read error: %s", e)
        finally:
            self.connected = False
            self._dispatch_disconnect(reason)


class WebSocketTransport(Transport):
    def __init__(self, url="ws://localhost:8080/kernel"):
        super().__init__()
        self.url = url
        self._ws = None
        self._read_task: asyncio.Task | None = None

    async def connect(self):
        import websockets
        self._ws = await websockets.connect(self.url)
        self.connected = True
        self._read_task = asyncio.create_task(self._read_loop())
        log.info("websocket connected to %s", self.url)

    async def disconnect(self):
        self.connected = False
        if self._read_task:
            self._read_task.cancel()
            try:
                await self._read_task
            except asyncio.CancelledError:
                pass
        if self._ws:
            await self._ws.close()
            self._ws = None
        log.info("websocket disconnected")

    async def send(self, msg: dict) -> bool:
        if not self._ws:
            return False
        try:
            await self._ws.send(self._frame_encode(msg))
            return True
        except Exception as e:
            log.error("websocket send failed: %s", e)
            return False

    async def _read_loop(self):
        buf = b""
        reason = "websocket eof"
        try:
            async for data in self._ws:
                if isinstance(data, str):
                    data = data.encode()
                buf += data
                while b"\x00" in buf:
                    idx = buf.index(b"\x00")
                    frame = buf[:idx]
                    buf = buf[idx + 1:]
                    if not frame:
                        continue
                    try:
                        msg = self._frame_decode(frame)
                        self._dispatch(msg)
                    except Exception as e:
                        log.error("frame decode error: %s", e)
        except asyncio.CancelledError:
            reason = "websocket closed"
        except Exception as e:
            reason = f"websocket read error: {e}"
            if self.connected:
                log.error("websocket read error: %s", e)
        finally:
            self.connected = False
            self._dispatch_disconnect(reason)


def create_transport(transport_type: str, **kwargs) -> Transport:
    t = transport_type.lower()
    if t in ("tcp", "socket"):
        return TCPTransport(
            host=kwargs.get("host", "localhost"),
            port=kwargs.get("port", 5555),
        )
    if t == "serial":
        return SerialTransport(
            port_path=kwargs.get("port_path", "/dev/ttyUSB0"),
            baud_rate=kwargs.get("baud_rate", 115200),
        )
    if t in ("websocket", "ws"):
        return WebSocketTransport(
            url=kwargs.get("url", "ws://localhost:8080/kernel"),
        )
    raise ValueError(f"unknown transport type: {transport_type}")
